flink-kubernetes-operator-autoscaler/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/RecommendedParallelismTest.java

```diff
@@ -0,0 +1,296 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.operator.autoscaler;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.kubernetes.operator.OperatorTestBase;
+import org.apache.flink.kubernetes.operator.TestUtils;
+import org.apache.flink.kubernetes.operator.api.FlinkDeployment;
+import org.apache.flink.kubernetes.operator.autoscaler.config.AutoScalerOptions;
+import org.apache.flink.kubernetes.operator.autoscaler.metrics.FlinkMetric;
+import org.apache.flink.kubernetes.operator.autoscaler.metrics.ScalingMetric;
+import org.apache.flink.kubernetes.operator.autoscaler.topology.JobTopology;
+import org.apache.flink.kubernetes.operator.autoscaler.topology.VertexInfo;
+import org.apache.flink.kubernetes.operator.config.FlinkConfigManager;
+import org.apache.flink.kubernetes.operator.reconciler.ReconciliationUtils;
+import org.apache.flink.kubernetes.operator.utils.EventCollector;
+import org.apache.flink.kubernetes.operator.utils.EventRecorder;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.rest.messages.job.metrics.AggregatedMetric;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.server.mock.EnableKubernetesMockClient;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import lombok.Getter;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.kubernetes.operator.autoscaler.metrics.ScalingMetric.PARALLELISM;
+import static org.apache.flink.kubernetes.operator.autoscaler.metrics.ScalingMetric.RECOMMENDED_PARALLELISM;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+/** Test for recommended parallelism. */
+@EnableKubernetesMockClient(crud = true)
+public class RecommendedParallelismTest extends OperatorTestBase {
```

Comment:
I generally like the idea of having a separate test but this basically duplicates one of our worst / hardest-to-maintain tests that we wanted to rethink over time.

Would it be possible to just add the metric assertions to the existing tests? I know it is not ideal but maybe that will cause less pain over time.

Reply:
The problem with these tests are that they tend to break when we touch the scaling logic / default configs. So now instead of one test breaking and hard 2 fix, we have 2 tests that can break and are hard to fix

Reply:
I agree with both of you that (a) having a dedicated test is preferable (b) duplicating most of the logic of an already hard to maintain test should be avoided. I think a compromise would be to add assertions to the existing end to end test and add a dedicated unit/integration test which tests only the core assumptions without going through the entire lifecycle.

Reply:
I agree with both of you that (a) having a dedicated test is preferable (b) duplicating most of the logic of an already hard to maintain test should be avoided. I think a compromise would be to add assertions to the existing end to end test and add a dedicated unit/integration test which tests only the core assumptions without going through the entire lifecycle. 

Adding a test method to ScalingExecutor should be sufficient IMHO to test that the recommended parallelism is added. The cleanup can be tested in the end to end test. If we want to go all the way, we could test the cleanup in JobAutoscalerImpl without the entire end to end setup, e.g. look at JobAutoscalerImplTest#testErrorReporting which only tests the error reporting. Since you are cleaning up in the finally block, the test setup should work similarly.

Reply:
I discussed this with Gyula and we agreed the e2e for the recommended parallelism can stay and will be refactored together with the backlog e2e.

Reply:
Why did we not consider any of the proposed alternatives? What is the point of a review if no discussion takes place?

Reply:
@mxm I discussed this with Matyas and he prefered to keep this functionality independently and comprehensively tested instead of adding bits and pieces to other tests. This functionality involves the full lifecycle of the scaling logic and the test makes sure that the recommendation is conistsent along the way.

As you (@mxm) and I discussed we should come up with a way to have better and more robust end to end tests but this is clearly out of scope here. I support the decision made here, as I feel the feedback was considered fairly :)

Reply:
There is no point in taking the time to review and make suggestions if those are dismissed after internal discussions. I would appreciate if we had those discussions here. 

I think we need to consider what the right amount of test coverage would be. There isn't a single unit or integration test here. We cannot add an e2e test for every feature. IMHO the impact of this feature does not justify copying questionable end to end testing code. This could have been tested in a more light way with comparable test coverage.

Reply:
@mxm I think thats fair point. The review wasn't ignored in this case but you are completely right that no appropriate feedback was provided before going ahead with this change.

@morhidi I think the best course of action would be to work on simplifying the test and adopt a more unit test style as Max suggested. @morhidi could you please work on that as a separate PR?

Reply:
Thanks @mxm for your comprehensive review and feedbacks. I agree we can/should be more careful with tracking internal decisions upstream. I'll follow up on that @gyfora thanks.